principles/coding-principles.md

```diff
@@ -89,3 +89,45 @@ Don’t swallow unexpected open type or base type exceptions in your libraries,
 ### 2. PEP 8
 Always try to follow [pep8 code style](https://www.python.org/dev/peps/pep-0008/) and
 consider the [Zen of Python](https://www.python.org/dev/peps/pep-0020/)
+
+## Testing-specific principles
+
+### 1. Write tests
+Always write tests for any code that you write. Besides giving developers
+confidence that the code works, the tests will give context that otherwise
+would be missing. This is especially valuable a few years down the road when
+all of the original developers have left the project.
+
+### 2. Use test-driven development or behaviour-driven development
+Use test-driven development or behaviour-driven development if possible. It
+ensures that before you write your code that you have success criteria, and
+that your code is testable.
+
+### 3. Follow the testing pyramid
+The [testing pyramid](https://martinfowler.com/bliki/TestPyramid.html) makes
+the point to write many low level unit tests, fewer integration tests, and even
+fewer acceptance tests.
+
+### 4. Decide on a testing strategy
+At the beginning of a new project, it's good to decide on a testing strategy as
+a team. Decide what kind of testing tools to use and what types of tests to
+write, using the testing pyramid as the basis.
+
+### 5. Mock consistently
+When doing TDD, developers tend to follow either the [London or the Chicago](https://softwareengineering.stackexchange.com/questions/123627/what-are-the-london-and-chicago-schools-of-tdd)
+school of TDD.
+
+If you follow the London (mockist) style of TDD, make sure you
+have sufficient integration tests to compensate for all the mocking.
+
+Decide as a team which style you prefer and use it consistently across the
+project, rather than mixing the two.
+
+### 6. Measure test coverage
+Measure the test coverage of your project and make your build fail if it is not
+at 100%. If you cannot do that, then make your build fail if the test coverage
```

Comment:
IME striving for 100% can lead to testing for the sake of it. Lightly suggest 90% as a more reasonable coverage number.

Reply:
I stick by the 100% for the reason that tests = documentation. Having legacy code without tests can be a massive pain when you wonder why something is the way it is.

Reply:
Taking liberally from [Martin Fowler](https://www.martinfowler.com/bliki/TestCoverage.html), I've seen attempts to enforce 100% code coverage leading to low-quality tests, as people naturally optimise for KPIs. To quote the article directly:

> I would be suspicious of anything like 100% - it would smell of someone writing tests to make the coverage numbers happy, but not thinking about what they are doing.

Jumping back to the opening comment of this PR:

> The goal is to ensure good test quality and coverage, without taking away the freedom to write tests in a way that the team is happy with.

This feels like it goes a little against that by prescribing a goal of 100%. Echoing @tvararu a little, perhaps having a note about the team agreeing their code coverage targets, with a lower limit of 85-90% might be good.

Reply:
> I would be suspicious of anything like 100% - it would smell of someone writing tests to make the coverage numbers happy, but not thinking about what they are doing.

I've only worked on projects with 100% coverage in the past years, or 99+ at least, this was never the case. I don't even know how you actually accomplish what is described in the article, using our tools. Perhaps that is why I don't understand where you are coming from.

The only feasible way of getting false positive test coverage would be to have a single (or few) feature/integration test that manages to touch everything, which is why I recommend not having them included in coverage checks.

This might be different from other stacks than Ruby however, I don't know. These guidelines need to work across other stacks.

Reply:
Besides, these are guidelines, not rules :)

Reply:
I've adjusted the guideline for test coverage now as I understand 100% coverage in other stacks in troublesome.